src/ray/raylet/lineage_cache.cc

```diff
@@ -297,42 +330,77 @@ void LineageCache::Flush() {
   }
 }
 
-void PopAncestorTasks(const UniqueID &task_id, Lineage &lineage) {
-  auto entry = lineage.PopEntry(task_id);
+bool LineageCache::SubscribeTask(const UniqueID &task_id) {
+  auto inserted = subscribed_tasks_.insert(task_id);
+  if (inserted.second) {
+    // Request notifications for the task if we haven't already requested
+    // notifications for it.
+    RAY_CHECK_OK(task_pubsub_.RequestNotifications(JobID::nil(), task_id, client_id_));
+    return true;
+  } else {
+    return false;
+  }
+}
+
+bool LineageCache::UnsubscribeTask(const UniqueID &task_id) {
+  auto it = subscribed_tasks_.find(task_id);
+  if (it != subscribed_tasks_.end()) {
+    // Cancel notifications for the task if we previously requested
+    // notifications for it.
+    RAY_CHECK_OK(task_pubsub_.CancelNotifications(JobID::nil(), task_id, client_id_));
+    subscribed_tasks_.erase(it);
+    return true;
+  } else {
+    return false;
+  }
+}
```

Comment:
Purely a style thing, but I think it'd be cleaner to have a single return statement and to do something like

```c++
bool task_present = it != subscribed_tasks_.end();
if (task_present) {
  ...
}
reeturn task_present;
```

and similarly for the above block.